sympy/physics/mechanics/functions.py

```diff
@@ -532,7 +539,7 @@ def _recurser(expr, sub_dict):
 
 def _fraction_decomp(expr):
     """Return num, den such that expr = num/den"""
-    if not isinstance(expr, Mul):
+    if not isinstance(expr, MulClass):
```

Comment:
@asmeurer I think the compatibility layer has nothing to do with C or C++, it's just how the Python wrappers are arranged. I agree that the
```
if HAS_SYMENGINE:
    ...
else:
    ...
```
should only be in the compatibility layer, and once you import `sin`, it needs to be 100% SymPy compatible (no matter how it's implemented underneath). The method described in https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine will just work and always fallback to sympy. It's just that it will constantly convert between sympy and symengine for each operation, i.e. if you call a.diff(x), where "a" and "x" are sympy objects, it will convert them to symengine, do the diff (in C++) and convert back to sympy. You can of course instead return the symengine Python expression from a.diff(x), but then this symengine Python object must have all the methods and behave like a sympy object. Then of course, `.args` must return sympy compatible objects and so on.

One motivation behind https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine is that it's slow, but ultimately, it will eventually call symengine for all operations that symengine can do. So you can literally remove most of sympy.core (if we decide to only use symengine --- which I think we shouldn't, see below), 
and then you can remove the conversions to and from sympy. Only at this point things will get fast, since no more conversions are done. One can then speed things up by calling into symengine.py directly, bypassing the compatibility layer.

The problem is that this assume we'll remove sympy.core. I wouldn't, I would always keep it around, I think it's good to have a pure Python implementation. So I think we need a different solution, one that allows us to incrementally use symengine and be fast from the very beginning, yet compatible with sympy.core.

The advantage is that the symengine/symengine repo has no Python, so we make sure that one is fast. The Python wrappers are slow just because we use Python. So if we make it a bit more slower by being compatible with sympy, I think we actually don't care much about that, because if there is something that needs to be fast, we simply implement it in C++, make it fast, and then just wrap it into Python.

One solution would be to have two wrappers ---- one set of wrappers that are as fast as one can get in Cython, but not compatible with sympy. The other wrappers would be 100% compatible with sympy, but slower. But it's a lot of work to maintain two sets of wrappers.

And so perhaps we can just make the wrappers themselves 100% compatible with sympy.

Ok, maybe I found a solution. I was always imagining the wrappers to have sympy objects, convert to symengine, call C++, and convert back to sympy. However, we should go the other way around --- each class like `sin` would implement all the methods of the sympy's sin, and each method would accept either sympy or symengine objects, convert to sympy, call sympy, and convert to symengine and return the result. So `sin(x).diff(x)` would return symengine's `cos(x)`. That way, if the `sin(x)` was already a symengine object, then no conversion is taking place and things are fast. At the same time, this `cos(x)` that it returns is a symengine object, that will however implement all sympy's methods, and so for the rest of sympy, it would be a drop in replacement.

And so that way, I think we can actually make symengine a drop in replacement for sympy objects, step by step. And things will still be fast, as long as the user keeps using symengine objects.

But here is another problem. Say complex numbers --- for speed reasons, `1+I` is a Complex class in C++. It is an `Add(1, I)` in SymPy. I don't think there is a way to make this 100% compatible. In other words, if `a.diff(x)` returns `1+I`, it will be a Complex class. SymPy doesn't have such a class. So things will break. I don't have a solution to this problem.

The other way, i.e. keeping things as sympy objects by default, as described here:

https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine

doesn't have this problem, since we never return a symengine object, we always convert it first to sympy, so a Complex class `1+I` becomes SymPy's `Add(1,I)`. So that's why I wrote this up in the wiki, as I know this will deliver. But this approach will necessarily be slower, since there are these conversions back and forth. Only at the very end, after all of sympy uses the compatibility layer and never sympy.core directly, only then one can remove the conversions back and forth, and things become fast. But that's a lot of work, and in the meantime, things might even become slower when you turn on symengine, since for simple things like `x+x` you have to convert back and forth all the time.

This PR is a middle ground. It doesn't try to be 100% compatible, it doesn't do any conversions and returns symengine objects. Thus it is fast.

Anyway, I don't know the best way forward.